### DISCUSSION TITLE
Approach to add capability to highlighter without changing codebase

### DISCUSSION THREAD

**Discussion by mdgilger** [stephanrauh/ngx-extended-pdf-viewer#2338] [Ideas]
This library is great! Was using the ngx-pdf-viewer, but the necessary functionality we desired was too much code.  ngx-extended-pdf-viewer is a better fit for what we require, but we do want to add in some functionality that isn't currently in the set of events that we can intercept.  In this case, it is the highlighter capability. A great capability, and we would like to extend it in a couple of ways. 

1) Two parts to this one: a) Would like the button on the tool bar to be a "mode" button - so you press it, and it indents to show the user that it is in markup mode, because b) we don't want the dialog that currently displays that allows them to choose the color, thickness, etc.  
2) When you complete a highlight, or select one by clicking on it - love the pop-up that comes up as far as a GUI construct. However, we want to display different things in that popup (a couple of buttons that will display popup dialogs). Was thinking that through eventing, we could do this without modifying the base source code (providing a "highlight selected" and a "highlight completed" event so that we could display our own dialog.

We are willing to do this work, but want to make sure that we do it in a way that enhances ngx-extended-pdf-viewer in a way that is acceptable (so you will except the pull requests :-) ), and fits the design patterns used in the library.  Can you outline the approach you would want us to take to bring this capability in, please?

Note, we will be adding tags and notes - which we would like to share - but we won't modify the original PDF - in our customer base, that is forbidden - so we create a JSON that contains all of the info - and pass it to services that persist it, linking it back to the original PDF, but these services can be modified so that you can tie it into any kind of backend system to story this info.

Input welcome!  Thanks!

**Comment by mdgilger** [comment]
Perhaps the title isn't completely accurate - thinking that we would initially need to modify the code base to support these new events, but the actual goal is to make it so that with small mods to the code base - others will be able to intercept and add capabilities going forward without changing the code base.

**Comment by stephanrauh** [comment]
Your ideas make sense to me, but I'm afraid it might be a bit difficult to implement them, so I'm sort of happy you've closed the ticket because I had an opportunity to read it. :)

In any case, if the events "highlight selected" and "highlight completed" aren't there yet, that's a bug. Feel free to open a feature request if these events aren't covered by `(annotationEditorEvent)` yet.
